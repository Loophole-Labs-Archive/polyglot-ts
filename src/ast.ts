/*
	Copyright 2022 Loophole Labs

	Licensed under the Apache License, Version 2.0 (the "License");
	you may not use this file except in compliance with the License.
	You may obtain a copy of the License at

		   http://www.apache.org/licenses/LICENSE-2.0

	Unless required by applicable law or agreed to in writing, software
	distributed under the License is distributed on an "AS IS" BASIS,
	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
	See the License for the specific language governing permissions and
	limitations under the License.
*/

interface ISrcAST {
  [k: string]: protobuf.AnyNestedObject;
}

interface ISrcType {
  fields: {
    [k: string]: ISrcField;
  };
  values: {
    [k: string]: number;
  };
}

interface ISrcField {
  type: string;
  id: number;
  rule: string;
  keyType: string;
}

interface IDstType {
  typeName: string;
  fields: IDstField[];
}

interface IDstField {
  fieldName: string;
  typeName: string;
  keyTypeName: string;
  isArray: boolean;
  isMap: boolean;
}

interface IDstEnum {
  enumName: string;
  values: string[];
}

export const getTypes = (root: ISrcAST): IDstType[] =>
  Object.keys(root)
    .filter((typeName) => (root[typeName] as ISrcType).fields)
    .map((typeName) => {
      const srcTypeFields = (root[typeName] as ISrcType).fields;

      return {
        typeName,
        fields: Object.keys(srcTypeFields)
          .map((fieldName) => ({
            fieldName,
            typeName: srcTypeFields[fieldName].type,
            keyTypeName: srcTypeFields[fieldName].keyType,
            index: srcTypeFields[fieldName].id,
            isArray: srcTypeFields[fieldName].rule === "repeated",
            isMap: !!srcTypeFields[fieldName].keyType,
          }))
          .sort((curr, prev) => curr.index - prev.index)
          .map((field) => ({
            fieldName: field.fieldName,
            typeName: field.typeName,
            keyTypeName: field.keyTypeName,
            isArray: field.isArray,
            isMap: field.isMap,
          })),
      };
    });

export const getEnums = (root: ISrcAST): IDstEnum[] =>
  Object.keys(root)
    .filter((typeName) => (root[typeName] as ISrcType).values)
    .map((typeName) => {
      const srcTypeFields = (root[typeName] as ISrcType).values;

      return {
        enumName: typeName,
        values: Object.keys(srcTypeFields)
          .map((field) => ({
            key: field,
            value: srcTypeFields[field],
          }))
          .sort((a, b) => a.value - b.value)
          .map((value) => value.key),
      };
    });

export const getRootAndNamespace = (
  obj: any,
  objPath: string[] = [],
  lastObj: any = undefined
): { root: any; namespace: string[] } => {
  if (!obj || obj.fields || obj.methods) {
    return {
      root: lastObj,
      namespace: objPath.slice(0, -1),
    };
  }

  let nextKey = Object.keys(obj)[0];
  if (nextKey === "options") {
    nextKey = Object.keys(obj)[1];
  }

  if (nextKey === "nested") {
    return getRootAndNamespace(obj[nextKey], objPath, obj);
  }

  return getRootAndNamespace(obj[nextKey], [...objPath, nextKey], obj);
};
