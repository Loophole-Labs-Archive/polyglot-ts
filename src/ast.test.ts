/*
	Copyright 2022 Loophole Labs

	Licensed under the Apache License, Version 2.0 (the "License");
	you may not use this file except in compliance with the License.
	You may obtain a copy of the License at

		   http://www.apache.org/licenses/LICENSE-2.0

	Unless required by applicable law or agreed to in writing, software
	distributed under the License is distributed on an "AS IS" BASIS,
	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
	See the License for the specific language governing permissions and
	limitations under the License.
*/

import path from "path";
import protobuf from "protobufjs";
import { getEnums, getRootAndNamespace, getTypes } from "./ast";

describe("AST", () => {
  it("Can get root and namespace from simple protobuf", () => {
    const proto = protobuf
      .loadSync(path.join(__dirname, "..", "data", "simple.proto"))
      .toJSON().nested!;

    const { root, namespace } = getRootAndNamespace(proto);

    expect(namespace).toEqual([
      "io",
      "loopholelabs",
      "polyglot",
      "test",
      "data",
      "simple",
    ]);
    expect(root).toMatchSnapshot();
  });

  it("Can get AST from simple protobuf", () => {
    const proto = protobuf
      .loadSync(path.join(__dirname, "..", "data", "simple.proto"))
      .toJSON().nested!;

    const { root } = getRootAndNamespace(proto);
    const types = getTypes(root);

    expect(types).toMatchSnapshot();
  });

  it("Can get root and namespace from complex, unordered protobuf", () => {
    const proto = protobuf
      .loadSync(path.join(__dirname, "..", "data", "complex.proto"))
      .toJSON().nested!;

    const { root, namespace } = getRootAndNamespace(proto);

    expect(namespace).toEqual([
      "io",
      "loopholelabs",
      "polyglot",
      "test",
      "data",
      "complex",
    ]);
    expect(root).toMatchSnapshot();
  });

  it("Can get AST from complex, unordered protobuf", () => {
    const proto = protobuf
      .loadSync(path.join(__dirname, "..", "data", "complex.proto"))
      .toJSON().nested!;

    const { root } = getRootAndNamespace(proto);

    const types = getTypes(root);
    expect(types).toMatchSnapshot();

    const enums = getEnums(root);
    expect(enums).toMatchSnapshot();
  });

  it("Can get AST from protobuf with array", () => {
    const proto = protobuf
      .loadSync(path.join(__dirname, "..", "data", "array.proto"))
      .toJSON().nested!;

    const { root } = getRootAndNamespace(proto);

    const types = getTypes(root);
    expect(types).toMatchSnapshot();

    const enums = getEnums(root);
    expect(enums).toMatchSnapshot();
  });

  it("Can get AST from protobuf with map", () => {
    const proto = protobuf
      .loadSync(path.join(__dirname, "..", "data", "map.proto"))
      .toJSON().nested!;

    const { root } = getRootAndNamespace(proto);

    const types = getTypes(root);
    expect(types).toMatchSnapshot();

    const enums = getEnums(root);
    expect(enums).toMatchSnapshot();
  });

  it("Can get AST from protobuf with enum", () => {
    const proto = protobuf
      .loadSync(path.join(__dirname, "..", "data", "enum.proto"))
      .toJSON().nested!;

    const { root } = getRootAndNamespace(proto);

    const types = getTypes(root);
    expect(types).toMatchSnapshot();

    const enums = getEnums(root);
    expect(enums).toMatchSnapshot();
  });
});
