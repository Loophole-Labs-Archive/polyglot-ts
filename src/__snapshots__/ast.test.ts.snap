// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AST Can get AST from complex, unordered protobuf 1`] = `
Array [
  Object {
    "fields": Array [
      Object {
        "fieldName": "a",
        "isArray": false,
        "isMap": false,
        "keyTypeName": undefined,
        "typeName": "double",
      },
      Object {
        "fieldName": "b",
        "isArray": false,
        "isMap": false,
        "keyTypeName": undefined,
        "typeName": "float",
      },
      Object {
        "fieldName": "c",
        "isArray": false,
        "isMap": false,
        "keyTypeName": undefined,
        "typeName": "int32",
      },
      Object {
        "fieldName": "d",
        "isArray": false,
        "isMap": false,
        "keyTypeName": undefined,
        "typeName": "int64",
      },
      Object {
        "fieldName": "e",
        "isArray": false,
        "isMap": false,
        "keyTypeName": undefined,
        "typeName": "uint32",
      },
      Object {
        "fieldName": "f",
        "isArray": false,
        "isMap": false,
        "keyTypeName": undefined,
        "typeName": "uint64",
      },
      Object {
        "fieldName": "g",
        "isArray": false,
        "isMap": false,
        "keyTypeName": undefined,
        "typeName": "sint32",
      },
      Object {
        "fieldName": "h",
        "isArray": false,
        "isMap": false,
        "keyTypeName": undefined,
        "typeName": "sint64",
      },
      Object {
        "fieldName": "i",
        "isArray": false,
        "isMap": false,
        "keyTypeName": undefined,
        "typeName": "fixed32",
      },
      Object {
        "fieldName": "j",
        "isArray": false,
        "isMap": false,
        "keyTypeName": undefined,
        "typeName": "fixed64",
      },
      Object {
        "fieldName": "k",
        "isArray": false,
        "isMap": false,
        "keyTypeName": undefined,
        "typeName": "sfixed32",
      },
      Object {
        "fieldName": "l",
        "isArray": false,
        "isMap": false,
        "keyTypeName": undefined,
        "typeName": "sfixed64",
      },
      Object {
        "fieldName": "m",
        "isArray": false,
        "isMap": false,
        "keyTypeName": undefined,
        "typeName": "bool",
      },
      Object {
        "fieldName": "n",
        "isArray": false,
        "isMap": false,
        "keyTypeName": undefined,
        "typeName": "string",
      },
      Object {
        "fieldName": "o",
        "isArray": false,
        "isMap": false,
        "keyTypeName": undefined,
        "typeName": "bytes",
      },
    ],
    "typeName": "ExampleMessage",
  },
]
`;

exports[`AST Can get AST from complex, unordered protobuf 2`] = `Array []`;

exports[`AST Can get AST from protobuf with array 1`] = `
Array [
  Object {
    "fields": Array [
      Object {
        "fieldName": "name",
        "isArray": false,
        "isMap": false,
        "keyTypeName": undefined,
        "typeName": "string",
      },
      Object {
        "fieldName": "addresses",
        "isArray": true,
        "isMap": false,
        "keyTypeName": undefined,
        "typeName": "string",
      },
      Object {
        "fieldName": "balances",
        "isArray": true,
        "isMap": false,
        "keyTypeName": undefined,
        "typeName": "int",
      },
    ],
    "typeName": "ExampleMessage",
  },
]
`;

exports[`AST Can get AST from protobuf with array 2`] = `Array []`;

exports[`AST Can get AST from protobuf with enum 1`] = `
Array [
  Object {
    "fields": Array [
      Object {
        "fieldName": "type",
        "isArray": false,
        "isMap": false,
        "keyTypeName": undefined,
        "typeName": "MessageType",
      },
    ],
    "typeName": "ExampleMessage",
  },
]
`;

exports[`AST Can get AST from protobuf with enum 2`] = `
Array [
  Object {
    "enumName": "MessageType",
    "values": Array [
      "MESSAGE_TYPE_A",
      "MESSAGE_TYPE_B",
      "MESSAGE_TYPE_C",
    ],
  },
]
`;

exports[`AST Can get AST from protobuf with map 1`] = `
Array [
  Object {
    "fields": Array [
      Object {
        "fieldName": "name",
        "isArray": false,
        "isMap": false,
        "keyTypeName": undefined,
        "typeName": "string",
      },
      Object {
        "fieldName": "balances",
        "isArray": false,
        "isMap": true,
        "keyTypeName": "string",
        "typeName": "int",
      },
      Object {
        "fieldName": "backlinks",
        "isArray": false,
        "isMap": true,
        "keyTypeName": "string",
        "typeName": "double",
      },
    ],
    "typeName": "ExampleMessage",
  },
]
`;

exports[`AST Can get AST from protobuf with map 2`] = `Array []`;

exports[`AST Can get AST from simple protobuf 1`] = `
Array [
  Object {
    "fields": Array [
      Object {
        "fieldName": "name",
        "isArray": false,
        "isMap": false,
        "keyTypeName": undefined,
        "typeName": "string",
      },
      Object {
        "fieldName": "age",
        "isArray": false,
        "isMap": false,
        "keyTypeName": undefined,
        "typeName": "int32",
      },
      Object {
        "fieldName": "department",
        "isArray": false,
        "isMap": false,
        "keyTypeName": undefined,
        "typeName": "Department",
      },
    ],
    "typeName": "User",
  },
  Object {
    "fields": Array [
      Object {
        "fieldName": "name",
        "isArray": false,
        "isMap": false,
        "keyTypeName": undefined,
        "typeName": "string",
      },
    ],
    "typeName": "Department",
  },
]
`;

exports[`AST Can get root and namespace from complex, unordered protobuf 1`] = `
Object {
  "ExampleMessage": Object {
    "fields": Object {
      "a": Object {
        "id": 1,
        "type": "double",
      },
      "b": Object {
        "id": 2,
        "type": "float",
      },
      "c": Object {
        "id": 3,
        "type": "int32",
      },
      "d": Object {
        "id": 4,
        "type": "int64",
      },
      "e": Object {
        "id": 5,
        "type": "uint32",
      },
      "f": Object {
        "id": 6,
        "type": "uint64",
      },
      "g": Object {
        "id": 7,
        "type": "sint32",
      },
      "h": Object {
        "id": 8,
        "type": "sint64",
      },
      "i": Object {
        "id": 9,
        "type": "fixed32",
      },
      "j": Object {
        "id": 10,
        "type": "fixed64",
      },
      "k": Object {
        "id": 11,
        "type": "sfixed32",
      },
      "l": Object {
        "id": 12,
        "type": "sfixed64",
      },
      "m": Object {
        "id": 13,
        "type": "bool",
      },
      "n": Object {
        "id": 14,
        "type": "string",
      },
      "o": Object {
        "id": 15,
        "type": "bytes",
      },
    },
  },
}
`;

exports[`AST Can get root and namespace from simple protobuf 1`] = `
Object {
  "Department": Object {
    "fields": Object {
      "name": Object {
        "id": 1,
        "type": "string",
      },
    },
  },
  "User": Object {
    "fields": Object {
      "age": Object {
        "id": 2,
        "type": "int32",
      },
      "department": Object {
        "id": 3,
        "type": "Department",
      },
      "name": Object {
        "id": 1,
        "type": "string",
      },
    },
  },
}
`;
